scripts/dev/check_constexpr.py

```diff
@@ -0,0 +1,431 @@
+    assert re.match(sci_notation_pattern, "-9.81e0")
+    assert re.match(sci_notation_pattern, "-9.81e-0")
+    assert re.match(sci_notation_pattern, "-9.81e+0")
+    assert re.match(sci_notation_pattern, "9.e0")
+    assert re.match(sci_notation_pattern, "9.e-0")
+    assert re.match(sci_notation_pattern, "9.e+0")
+    assert re.match(sci_notation_pattern, "+9.e0")
+    assert re.match(sci_notation_pattern, "+9.e-0")
+    assert re.match(sci_notation_pattern, "+9.e+0")
+    assert re.match(sci_notation_pattern, "-9.e0")
+    assert re.match(sci_notation_pattern, "-9.e-0")
+    assert re.match(sci_notation_pattern, "-9.e+0")
+
+    # don't match
+    assert not re.match(sci_notation_pattern, "VarName")
+
+
+def constexpr_check(search_path: Path):
+    errors_found = False
+
+    files_to_search = []
+    for p in [search_path]:
+        for root, dirs, files in os.walk(p):
+            for file in files:
+                f_path = Path(root) / Path(file)
+                f_extension = f_path.suffix
+                if f_extension == ".hh":
+                    files_to_search.append(f_path)
+                elif f_extension == ".cc":
+                    files_to_search.append(f_path)
+
+    files_to_search.sort()
+
+    for file in files_to_search:
+        with open(file, "r") as f:
+            lines = f.readlines()
+
+        lines = [x.strip() for x in lines]
+
+        bracket_count = 0
+
+        for idx, line in enumerate(lines):
+
+            # skip blank lines
+            if line == "":
+                continue
+
+            # skip comment lines
+            if line[0:2] == "//":
+                continue
+
+            # strip trailing comments
+            if "//" in line:
+                tokens = line.split("//")
+                line = tokens[0].strip()
+
+            if (re.match(const_num_pattern, line) or
+                re.match(array_const_pattern, line) or re.match(array_us_const_pattern, line)) and bracket_count == 0:
+                print("ERROR: 'const' found, convert to 'constexpr'")
+                s = f"{file.name}: {idx + 1} - {line}"
+                print(s)
+                errors_found = True
+
+            # count brackets in line
+            bracket_count += line.count("(")
+            bracket_count -= line.count(")")
+
+    return errors_found
+
+
+if __name__ == "__main__":
+
+    # run tests
+    if True:
+        test_array_underscore_const()
+        test_array_const()
+        test_const_num()
+        test_const_num_equal()
+        test_const_num_paren()
+        test_var_name()
+        test_const_type()
+        test_int_real()
+        test_sci_notation()
+
+    root_path = Path(__file__).parent.parent.parent
+    src_path = root_path / "src" / "EnergyPlus"
+    tst_path = root_path / "tst" / "EnergyPlus" / "unit"
+    errors_found = constexpr_check(src_path)
+    errors_found &= constexpr_check(tst_path)
+
+    if errors_found:
+        raise sys.exit(1)
```

Comment:
This is a really great addition.  I modified it locally to take advantage of Python unittest stuff, but did not change the functionality of it.  I'll push those changes with any other stuff I change or merges from develop.

Reply:
I had all using `unittest` originally but my system wouldn't let me run the tests and the actual checks together. 🤷  Fix'r up!